### PR TITLE
Fix loss of positive coinductive SR by restricting return clauses

### PR DESCRIPTION
**Pull request by ppedrot**
Implement a BTT-style restriction for positive coinductive types.

## Outline

The idea is to check that the return predicate of a coinductive pattern-matching is stable by cofix expansion. It is literally the one rule that has to hold for subject reduction to be preserved. The criterion is the following: assume w.l.o.g. an indexed coinductive type `T : forall i : I, Type`. A pattern-matching with return clause `P : forall i : I, T i -> Type` is valid iff it satisfies the following definitional equation:
```i : I, x : T i ⊢ P i x ≡ P i (cofix F := x).```

It is straightforward to see that this condition is necessary for SR to hold, and furthermore it can be shown to be sufficient as well, assuming a bit of nice meta-theoretical properties on the syntax of CIC.

Thus this is, I believe, the *right* criterion.

## Code changes

We introduce an option `Lax CoInductive Match` to deactivate the check, and activate it by default. Whether or not performing the check is decided on a by-case basis (pun intended), i.e. every coinductive pattern-matching remembers that it is allowed or not to break subject reduction. (This is necessary to preserve SR.) Eventually, this backward compatibility feature will be removed from the kernel, and thus the flag from the pattern-matching information as well.

This is an alternative to #7536. It seems that providing a good transition path was asked for in that PR, which is precisely what the current PR is achieving.

## Type-theoretical details

Let me explain intuitively where the criterion comes from, for people not versed in the semantics of type theory.

In Coq, you cannot unfold arbitrarily (co-)fixpoints, otherwise this leads to non-decidability of typechecking. In their infinite wisdom, the people who implemented (co-)inductive types came up with a syntactic guard for (co-)fixpoint unfolding, namely:
- for inductive fixpoints, they only unfold when applied to something that starts with a constructor, e.g. `(fix F := t) (S n) ≡ t{F := (fix F := t)} (S n)`
- for co-inductive cofixpoints, they only unfold when they are inside a pattern-matching, e.g.
`match (cofix F := t) with p ≡ match t{F := (cofix F := t)} with p`.

The latter corresponds intuitively to the fact that the cofix is being *forced*.

Now, assume some dependent pattern-matching over a term of coinductive type `T`. In general, this pattern-matching is of the shape `match t as x return P x with p`. Its type is thus `P t`. But now, assume that `t ≡ (cofix F := t₀)`, in which case the unfolding rule mandates that expression is convertible to `match t₀{F := (cofix F := t₀)} as x return P x with p`, which now has type `P (t₀{F := (cofix F := t₀)})`.

Therefore, for SR to hold we **need** to have `P (cofix F := t₀) ≡ P (t₀{F := (cofix F := t₀)})`.

Because `t₀` is arbitrary, this could very well be a variable as well. The criterion is based on this observation, and it turns out that this is actually sufficient for SR to hold in general.

## TODO

- Documentation and tests
- Adapt the pretyper and the `case` tactic (help needed!)

## Overlays
- https://gitlab.inria.fr/fpottier/menhir/-/merge_requests/30
- https://github.com/coq-community/paramcoq/pull/94
- https://github.com/charguer/tlc/pull/9

### CODE REVIEW

**Comment by ppedrot** [comment]
Summoning people that were discussing the other PR like @aspiwack, @gilhur, @JanBessai, @herbelin, @mattam82, @maximedenes, @xavierleroy.

**Comment by ppedrot** [comment]
I have an interesting problem from the test-suite though. I don't know whether the following is provable in this system:
```
CoInductive LList A :=
| LNil : LList A
| LCons : A -> LList A -> LList A.

Inductive Finite A : LList A -> Prop :=
| Lnil_fin : Finite A (LNil A)
| Lcons_fin : forall a l, Finite A l -> Finite A (LCons A a l).

CoInductive Infinite A : LList A -> Prop :=
| LCons_inf : forall a l, Infinite A l -> Infinite A (LCons A a l).

Lemma Not_Finite_Infinite : forall A (l : LList A), ~ Finite A l -> Infinite A l.
Proof.
...
```

**Comment by herbelin** [comment]
> `Lemma Not_Finite_Infinite : forall A (l : LList A), ~ Finite A l -> Infinite A l.`

I would suspect that we can prove this in current Coq using a proof such that for `l` be, say, a stream of zeros, the subject reduction issue will show up (i.e. the `destruct l` will reduce in the proof, but the `l` will not in the statement). Maybe by calling `LCons_inf` once in the proof before the `cofix` so that the `cofix` does not block the reduction of the `destruct l`.

**Comment by JasonGross** [comment]
> Maybe by calling `LCons_inf` once in the proof before the `cofix` so that the `cofix` does not block the reduction of the `destruct l`.

I don't think you get to do that, because you need `cofix` before the `intro`s so that the coinduction hypothesis is appropriately quantified.  At best, you can move it up to come before the inner `destruct`.

I tried to prove this with coinductive records.  I got a proof, but when I `Set Primitive Projections`, I get stuck without an eta axiom.
```coq
CoInductive LStream A := SCons { Hd : A ; Tl : option (LStream A) }.
Notation LList A := (option (LStream A)).
Definition LNil A : LList A := None.
Definition LCons A (a : A) (l : LList A) : LList A := Some (SCons A a l).

Inductive Finite A : LList A -> Prop :=
| Lnil_fin : Finite A (LNil A)
| Lcons_fin : forall a l, Finite A l -> Finite A (LCons A a l).

CoInductive Infinite A (ll : LList A) : Prop :=
  LCons_inf' { inf : exists a l, ll = LCons A a l /\ Infinite A l }.


Lemma Not_Finite_Infinite : forall A (l : LList A), ~ Finite A l -> Infinite A l.
Proof.
  cofix IH.
  intros A l H.
  constructor.
  destruct l; revgoals.
  { exfalso; apply H; constructor. }
  { destruct l as [a l].
    exists a, l; split; [ reflexivity | ].
    apply IH.
    intro H'; apply H.
    constructor; exact H'. }
Qed.
Print Not_Finite_Infinite.
(* Not_Finite_Infinite =
cofix IH (A : Type) (l : LList A) (H : ~ Finite A l) : Infinite A l :=
  {|
  inf := match
           l as o
           return (~ Finite A o -> exists (a : A) (l0 : LList A), o = LCons A a l0 /\ Infinite A l0)
         with
         | Some l0 =>
             fun H0 : ~ Finite A (Some l0) =>
             match
               l0 as l1
               return
                 (~ Finite A (Some l1) ->
                  exists (a : A) (l2 : LList A), Some l1 = LCons A a l2 /\ Infinite A l2)
             with
             | {| Hd := a; Tl := l1 |} =>
                 fun H1 : ~ Finite A (Some {| Hd := a; Tl := l1 |}) =>
                 ex_intro
                   (fun a0 : A =>
                    exists l2 : LList A,
                      Some {| Hd := a; Tl := l1 |} = LCons A a0 l2 /\ Infinite A l2) a
                   (ex_intro
                      (fun l2 : LList A =>
                       Some {| Hd := a; Tl := l1 |} = LCons A a l2 /\ Infinite A l2) l1
                      (conj eq_refl (IH A l1 (fun H' : Finite A l1 => H1 (Lcons_fin A a l1 H')))))
             end H0
         | None =>
             fun H0 : ~ Finite A None =>
             False_ind (exists (a : A) (l0 : LList A), None = LCons A a l0 /\ Infinite A l0)
               (H0 (Lnil_fin A))
         end H |}
     : forall (A : Type) (l : LList A), ~ Finite A l -> Infinite A l
*)
```
I suspect this is unprovable without eta.

**Comment by herbelin** [comment]
@JasonGross 

> I don't think you get to do that, because you need `cofix` before the `intro`s so that the coinduction hypothesis is appropriately quantified. At best, you can move it up to come before the inner `destruct`.

I think the following script works:
```
Set Implicit Arguments.

 CoInductive LList (A: Set) : Set :=
 |  LNil : LList A
 |  LCons : A -> LList A -> LList A.

Inductive Finite (A:Set) : LList A -> Prop :=
|  Lnil_fin : Finite (LNil A)
|  Lcons_fin : forall a l, Finite l -> Finite (LCons a l).

CoInductive Infinite  (A:Set) : LList A -> Prop :=
| LCons_inf : forall a l, Infinite l -> Infinite (LCons a l).

Lemma Not_Finite_Infinite : forall (A:Set)(l:LList A), ~ Finite l -> Infinite l.
Proof.
 destruct l.
  intro; absurd (Finite (LNil A));[auto|constructor].
  constructor. revert a l H.
 cofix H.
 destruct l.
 intro; absurd (Finite (LCons a (LNil A)));[auto|constructor]. apply Lnil_fin.
 constructor.
 eapply H.
 red; intro H1;case H0.
 constructor.
 eassumption.
Defined.

CoFixpoint S := LCons 1 S. 
Definition p := Not_Finite_Infinite (l:=S).
Compute p.
```
in the sense that it produces a proof of `Not_Finite_Infinite (l:=S)` which fails retype-checking with `cannot unify Finite (LCons 1 (cofix Fcofix : LList nat := LCons 1 Fcofix))` and `Finite S`.

So there seems indeed something crucial to do with eta in the validity of this statement which the (apparent) requirement to have the equality extensional in your "negative" definition of `Infinite` highlights. I suspect that one of the next 700 syntactical models of type theory would apply (??).

**Comment by xavierleroy** [comment]
I was summoned but I don't know what to say.  On the one hand it's nice to see an attempt to recover subject reduction that is less heavy-handed than "let's restrict coinduction to record types".  On the other hand, I can't get a good intuition on what the restriction on return clauses really means and whether it would be a problem in practice.  The `Not_Finite_Infinite` lemma is a good example of something that should remain provable.

**Comment by herbelin** [comment]
Hi @xavierleroy,
>  The `Not_Finite_Infinite` lemma is a good example of something that should remain provable.

My understanding is that the statement using the negative version of `Infinite` shall remain provable if the definition of `Infinite` is changed into:
```
CoInductive Infinite A (ll : LList A) : Prop :=
  LCons_inf' { inf : exists a l, EqSt ll (LCons A a l) /\ Infinite A l }.
```
for `EqSt` the extensional equality of streams.

**Comment by xavierleroy** [comment]
> [...] shall remain provable if the definition of Infinite is changed into [...]

Pardon my French, but this sucks. The definition of Infinite should not change.

For reference, here is the same development using a coinductive record to represent finite or infinite lists.  It is syntactically heavy but there is no need to define Infinite up to equivalence.
```
Set Implicit Arguments.

CoInductive LList (A: Set) : Set :=
  { this: option (A * LList A) }.

Definition LNil (A: Set) : LList A := {| this := None |}.
Definition LCons (A: Set) (hd: A) (tl: LList A) := {| this := Some(hd, tl) |}.

Inductive Finite (A:Set) : LList A -> Prop :=
|  Lnil_fin : forall l, l.(this) = None -> Finite l
|  Lcons_fin : forall l hd tl, l.(this) = Some(hd, tl) -> Finite tl -> Finite l.

CoInductive Infinite  (A:Set) : LList A -> Prop :=
| LCons_inf : forall l hd tl, l.(this) = Some(hd, tl) -> Infinite tl -> Infinite l.

Lemma Not_Finite_Infinite : forall (A:Set)(l:LList A), ~ Finite l -> Infinite l.
Proof.
  cofix COH. intros. destruct l as [[[hd tl] | ]].
- apply LCons_inf with hd tl. auto. apply COH. intros F; elim H. apply Lcons_fin with hd tl; auto.
- elim H. apply Lnil_fin; auto.
Qed.
```

**Comment by SkySkimmer** [comment]
>For reference, here is the same development using a coinductive record to represent finite or infinite lists. 

You need `Set Primitive Projections` to preserve SR, and it breaks the `destruct l as [[[hd tl] | ]].`

**Comment by SkySkimmer** [comment]
For `A = unit` `~ Finite -> Infinite` proves a something that looks a lot like the SR breaking theorem in https://github.com/coq/coq/issues/6768, so I'm not convinced that we're supposed to be able to prove it.

**Comment by xavierleroy** [comment]
> You need Set Primitive Projections to preserve SR, and it breaks the destruct l 

Oups!  Thanks for correcting me.  But pardon my French again: it sucks.  By this I mean that coinductive definitions that go beyond streams might very well become useless.

**Comment by xavierleroy** [comment]
Maybe my French was excessive, because here is a proof of `Not_Finite_Infinite` with records and primitive projections.
```
Set Implicit Arguments.
Set Primitive Projections.

CoInductive LList (A: Set) : Set :=
  { this: option (A * LList A) }.

Definition LNil (A: Set) : LList A := {| this := None |}.
Definition LCons (A: Set) (hd: A) (tl: LList A) := {| this := Some(hd, tl) |}.

Inductive Finite (A:Set) : LList A -> Prop :=
|  Lnil_fin : forall l, l.(this) = None -> Finite l
|  Lcons_fin : forall l hd tl, l.(this) = Some(hd, tl) -> Finite tl -> Finite l.

CoInductive Infinite  (A:Set) : LList A -> Prop :=
| LCons_inf : forall l hd tl, l.(this) = Some(hd, tl) -> Infinite tl -> Infinite l.

Lemma Not_Finite_Infinite : forall (A:Set)(l:LList A), ~ Finite l -> Infinite l.
Proof.
  cofix COH. intros. destruct l.(this) as [[hd tl] | ] eqn:E.
- apply LCons_inf with hd tl. auto. apply COH. intros F; elim H. apply Lcons_fin with hd tl; auto.
- elim H. apply Lnil_fin; auto.
Qed.
```

**Comment by ppedrot** [comment]
@xavierleroy in general you might need to massage a bit your definitions to keep the lemmas going through. Most notably, enforcing eta-expansability by hand in inductive predicates looks like a good design pattern. That said, positive coinduction is already fairly awkward in Coq, I don't think that the current PR makes things noticeably worse.

**Comment by ppedrot** [comment]
Also, let me explain intuitively where the criterion comes from, for people not versed in the semantics of type theory.

In Coq, you cannot unfold arbitrarily (co-)fixpoints, otherwise this leads to non-decidability of typechecking. In their infinite wisdom, the people who implemented (co-)inductive types came up with a syntactic guard for (co-)fixpoint unfolding, namely:
- for inductive fixpoints, they only unfold when applied to something that starts with a constructor, e.g. `(fix F := t) (S n) ≡ t{F := (fix F := t)} (S n)`
- for co-inductive cofixpoints, they only unfold when they are inside a pattern-matching, e.g.
`match (cofix F := t) with p ≡ match t{F := (cofix F := t)} with p`.

The latter corresponds intuitively to the fact that the cofix is being *forced*.

Now, assume some dependent pattern-matching over a term of coinductive type `T`. In general, this pattern-matching is of the shape `match t as x return P x with p`. Its type is thus `P t`. But now, assume that `t ≡ (cofix F := t₀)`, in which case the unfolding rule mandates that expression is convertible to `match t₀{F := (cofix F := t₀)} as x return P x with p`, which now has type `P (t₀{F := (cofix F := t₀)})`.

Therefore, for SR to hold we **need** to have `P (cofix F := t₀) ≡ P (t₀{F := (cofix F := t₀)})`.

Because `t₀` is arbitrary, this could very well be a variable as well. The criterion is based on this observation, and it turns out that this is actually sufficient for SR to hold in general.

TL; DR: dependent elimination over co-inductive types require a strictness criterion on their return predicate to ensure SR holds.

**Comment by xavierleroy** [comment]
@ppedrot: this is a nice explanation of the typing rule and of the "mechanics" of type-checking.  What humble users like me find difficult is to guess which statements of theorems fit the restriction and which do not.  Would you have guessed that `Not_Finite_Infinite` is problematic just by seeing its statement?

**Comment by xavierleroy** [comment]
>  in general you might need to massage a bit your definitions to keep the lemmas going through.  [...] That said, positive coinduction is already fairly awkward in Coq, I don't think that the current PR makes things noticeably worse.

I am not convinced at all.  Coq's positive coinduction leads to awkward __proofs__ (manual unrollings, nothing computes, etc) but the __specifications__ are quite nice.  You can take Capretta's PhD thesis, or the Leroy-Grall paper on coinductive natural semantics, and transcribe the mathematical descriptions as coinductive types and predicates that look like the maths.  

Here, you're complicating the definitions and arguing that the proofs get simpler.  This is a bad deal.  Nobody cares about the proofs.  But the definitions should look like maths and be crystal clear.  Turning sums-of-products into products-of-sums, baking in extensionality inside definitions, and whatever contorsions you will come up with next is not crystal clear and does not look like maths.

**Comment by JasonGross** [comment]
@xavierleroy I think I am starting to see a pattern here.  The return clause restriction says that you can never prove something about a coinductive by destructing it.  You are allowed to construct objects by destructing it (as long as the type of the object does not depend on the coinductive being  destructed), and you are allowed to prove things about the projections of a coinductive object by destructing the object.

**Comment by herbelin** [comment]
For the sake of completeness, here is an adaptation of @xavierleroy's idea to @ppedrot's initial question (thus using the positive-like formulation, compatible with the PR, and not using eta):
```
CoInductive LList (A: Set) : Set :=
 |  LNil : LList A
 |  LCons : A -> LList A -> LList A.
Arguments LNil {A}.
Arguments LCons {A}.

Inductive Finite {A} : LList A -> Prop :=
|  Lnil_fin : Finite LNil
|  Lcons_fin a l : Finite l -> Finite (LCons a l).

CoInductive Infinite {A} : LList A -> Prop :=
| LCons_inf a l : Infinite l -> Infinite (LCons a l).

Definition inspect {A} (l:LList A) :=
 match l with
 |  LNil => None
 |  LCons a l => Some (a,l)
 end.

Lemma Not_Finite_Infinite : forall A (l:LList A), ~ Finite l -> Infinite l.
Proof.
 cofix COH. intros * H. destruct (inspect l) as [(a,l')|] eqn:E.
 - assert (l = LCons a l') as ->.
   { destruct l. discriminate. injection E as [= -> ->]. trivial. }
   constructor.
   apply COH. intro. apply >H. constructor. trivial.
 - exfalso. apply >H. destruct l; [constructor | discriminate].
Defined.
```

**Comment by JasonGross** [comment]
@herbelin Does that actually build with this PR?  I expect that the `destruct l` used to prove `l = LCons a l'` should fail, because `l` is not guarded by a `match` on the LHS of the equality.

I think you genuinely need a different definition of `Infinite`, something like
```coq
CoInductive Infinite {A} : LList A -> Prop :=
| LCons_inf a l al : Infinite l -> inspect al = Some (a, l) -> Infinite al.
```

**Comment by ppedrot** [comment]
> Turning sums-of-products into products-of-sums, baking in extensionality inside definitions, and whatever contorsions you will come up with next is not crystal clear and does not look like maths.

@xavierleroy The same argument applies to function extensionality, and yet it seems Coq users have already gone beyond the five steps of grievance in that particular case. I don't see the difference with coinductive types, maybe that's just a matter of time?

**Comment by herbelin** [comment]
@ppedrot: very elegant _right_ criterion by the way!

**Comment by herbelin** [comment]
@JasonGross: you're right, I spoke too fast, sorry.

It works for the positive-like `LList` with the following variant of `Infinite`:
```
CoInductive Infinite {A} (l:LList A) : Prop :=
| LCons_inf a l' : inspect l = Some (a,l') -> Infinite l' -> Infinite l.

Lemma Not_Finite_Infinite : forall A (l:LList A), ~ Finite l -> Infinite l.
Proof.
 cofix COH. intros * H. destruct (inspect l) as [(a,l')|] eqn:E.
 - econstructor. rewrite E. eauto.
   apply COH. intro. apply H. destruct l; constructor. injection E as [= _ <-]. trivial.
 - exfalso. apply >H. destruct l; [constructor | discriminate].
Defined.
```
but I don't see how to invert `inspect l` without a dependent `destruct`. So, we are back to the  question of proving the unprovability in the absence of eta, sorry again.

As if `Infinite (cofix ...)` should hold for the proof to go through, but the definition of `Infinite` ending with `LCons` is demanding a too strong invariant (i.e. indeed demanding that an infinite `cofix` unfolds to an `LCons` even when not guarded).

**Comment by herbelin** [comment]
Actually, still not working: there are still `destruct` on `Finite`, so the definition of `Finite` should be changed also.

**Comment by herbelin** [comment]
> As if `Infinite (cofix ...)` should hold for the proof to go through, but the definition of `Infinite` ending with `LCons` is demanding a too strong invariant.

Applying the same reasoning to `Finite`, the following hopefully works w/o using a dependent 
case analysis in `LList`:
```
CoInductive LList (A: Set) : Set :=
 |  LNil : LList A
 |  LCons : A -> LList A -> LList A.
Arguments LNil {A}.
Arguments LCons {A}.

Definition inspect {A} (l:LList A) :=
 match l with
 |  LNil => None
 |  LCons a l => Some (a,l)
 end.

Inductive Finite {A} (l:LList A) : Prop :=
|  Lnil_fin : inspect l = None -> Finite l
|  Lcons_fin a l' : inspect l = Some (a,l') -> Finite l' -> Finite l.

CoInductive Infinite {A} (l:LList A) : Prop :=
| LCons_inf a l' : inspect l = Some (a,l') -> Infinite l' -> Infinite l.

Lemma Not_Finite_Infinite : forall A (l:LList A), ~ Finite l -> Infinite l.
Proof.
 cofix COH. intros * H. destruct (inspect l) as [(a,l')|] eqn:E.
 - econstructor. rewrite E. eauto.
   apply COH. intro. apply >H. apply Lcons_fin with a l'; auto.
 - exfalso. apply >H. apply Lnil_fin; auto.
Defined.
```
Sketching a syntactic impossibility result: for the statement to hold, it has to hold also for `cofix` expressions, but, with the initial definitions, this would require proving `~~ Finite cofix`, since there is no constructors in the definition of `Infinite` to prove `Infinite cofix`. For proving `~~ Finite cofix`, there is no alternatives than proving `Finite cofix` in the context `~ Finite cofix`, but again, there is no constructors to prove `Finite cofix`, so no normal proof can be found w/o actually destructing the `cofix`. (Maybe this was clear to @ppedrot from the very beginning...)

**Comment by herbelin** [comment]
Bouncing on @xavierleroy's [expectation](https://github.com/coq/coq/pull/10764#issuecomment-532696750) ("The definition of `Infinite` should not change") and @ppedrot's [comment](https://github.com/coq/coq/pull/10764#issuecomment-532827042) ("I don't see the difference [between accepting an axiom of function extensionality and accepting an axiom of coinductive extensionality]"), the somehow moral way (imo) to adapt to the removal (for decidability reasons) of the implicit definitional eta-expansion present in the historical dependent elimination rule for coinductive types would be to simply compensate by an explicit non-definitional (i.e. axiomatic) eta-rule, which in the given case would work as follows:
```
CoInductive LList (A: Set) : Set :=
 |  LNil : LList A
 |  LCons : A -> LList A -> LList A.
Arguments LNil {A}.
Arguments LCons {A}.

Inductive Finite {A} : LList A -> Prop :=
|  Lnil_fin : Finite LNil
|  Lcons_fin a l : Finite l -> Finite (LCons a l).

CoInductive Infinite {A} : LList A -> Prop :=
| LCons_inf a l : Infinite l -> Infinite (LCons a l).

Definition inspect {A} (l:LList A) :=
 match l with
 | LNil => None
 | LCons a l => Some (a,l)
 end.

Definition Build_LList {A:Set} (x:option (A * LList A)) : LList A :=
 match x with
 | None => LNil
 | Some (a,l) => LCons a l
 end.
 
Axiom eta_LList : forall {A} (l:LList A), l = Build_LList (inspect l).

Lemma Not_Finite_Infinite : forall A (l:LList A), ~ Finite l -> Infinite l.
Proof.
 cofix H.
 intros *.
 (* The next line is morally the only one to add *)
 (* to be compatible with the previous proof in the test-suite *)
 rewrite (eta_LList l).
 case l; intros. (* issue with "destruct" though, see below *)
 absurd (Finite (LNil (A:=A)));[auto|constructor].
 constructor.
 apply H.
 red; intro H1;case H0.
 constructor.
 trivial.
Defined.
```

As usual, the stronger statement of extensionality can be stated instead of eta:
```
CoInductive eqLList {A} (l1 l2:LList A) :=
| eqLList_LLnil :
  inspect l1 = None -> inspect l2 = None -> eqLList l1 l2 
| eqLList_LLcons a1 a2 l1' l2' :
  inspect l1 = Some (a1,l1') -> inspect l2 = Some (a2,l2') ->
  a1 = a2 -> eqLList l1' l2' -> eqLList l1 l2.

Axiom eqLList_extensional : forall {A} (l1 l2:LList A), eqLList l1 l2 -> l1 = l2.

Theorem eqLList_refl {A} : forall (l:LList A), eqLList l l.
Proof.
cofix refl. intro l.
  destruct (inspect l) as [(a,l')|] eqn:E.
  - apply eqLList_LLcons with a a l' l'; auto.
  - apply eqLList_LLnil; auto.
  Show Proof.
Defined.

Theorem eta_LList {A} (l:LList A) : l = Build_LList (inspect l).
Proof.
 apply eqLList_extensional.
 destruct (inspect l) as [(a,l')|] eqn:E.
  - apply eqLList_LLcons with a a l' l'; auto. apply eqLList_refl.
  - apply eqLList_LLnil; auto.
Defined.
```

Now, the next question is: can we compute with `eta_LList` or `eqLList_extensional`? Of course not in current Coq, but it would compute in a version of Coq with ad hoc polymorphic equality as can be found in observational type theory (case of a strict equality) or cubical type theory (case of a homotopic equality). In such a variant of Coq, equality on `LList` would be _per definition_ the extensional equality. The computational content for eta would simply be a coinductive proof reflecting the structure of the given `LList`, while the computational content of the extensionality axiom would trivially be the identity function. (One may then wonder how to rewrite with an extensional equality over streams: this is done by recursion on the structure of the proof of extensional equality, recursively rewriting the subcomponents of the stream.)

PS: In the proof above, I used `case` because the lax coinductive checker unexpectedly fails with `destruct` (it says: `Pattern-matching over a positive coinductive term has a non-strict return type "fun l : LList A0 => P l".` and it is not clear to me from where this `P` comes). Is this a bug in the PR or a weakness in `destruct`?

**Comment by ppedrot** [comment]
@herbelin the destruct error you mention is part of the reason why I am asking for help. For some reason, destruct generates a generic eliminator with an arbitrary predicate `P`, and somehow type-checks it before applying it to the goal to which the tactic is being applied. This leads to this puzzling error, and to fix it we would need destruct to know beforehand of the predicate that is being applied.

Note that I have the same kind of problem with the PR introducing the change of representation of pattern-matching. The destruct tactic is kind of a monstrous beast, and hardly amenable to low-level tweaks as it relies on the legacy engine.

**Comment by mattam82** [comment]
One more point for unifall which moves destruct to the new engine

**Comment by ppedrot** [comment]
@mattam82 what's the ETA of this part of unifall?

**Comment by herbelin** [comment]
@ppedrot, @mattam82: yes, that'd be good to know beforehand the predicate that is being applied and unifall will help. Tell me if there is something I can do to help.

**Comment by Zimmi48** [comment]
Dear PR author(s) and contributors, it seems to us this PR may need additional work beyond the 8.12 branching deadline on May 15th.

We are thus removing the 8.12+beta1 milestone; please retarget as appropriate, and also consider updating labels to reflect current status, moving the PR to draft status, etc.

We apologize in advance if we misunderstood the PR status.

**Comment by ppedrot** [comment]
Closing for now, I'll reopen when I have a backwards-compatible battle plan for this PR.
